crates/wit-parser/src/resolve.rs

```diff
@@ -158,69 +191,48 @@ impl Resolve {
-        let mut pkg = Some(pkg);
-        for name in order {
-            let pkg = deps.remove(&name).unwrap_or_else(|| pkg.take().unwrap());
-            files.extend(pkg.source_files().map(|p| p.to_path_buf()));
-            let pkgid = self.push(pkg)?;
-            last = Some(pkgid);
-        }
-
-        return Ok((last.unwrap(), files));
-
-        fn visit<'a>(
-            pkg: &'a UnresolvedPackage,
-            deps: &'a BTreeMap<PackageName, UnresolvedPackage>,
-            order: &mut IndexSet<PackageName>,
-            visiting: &mut HashSet<&'a PackageName>,
-        ) -> Result<()> {
-            if order.contains(&pkg.name) {
-                return Ok(());
-            }
-            pkg.source_map.rewrite_error(|| {
-                for (i, (dep, _)) in pkg.foreign_deps.iter().enumerate() {
-                    let span = pkg.foreign_dep_spans[i];
-                    if !visiting.insert(dep) {
-                        bail!(Error {
-                            span,
-                            msg: format!("package depends on itself"),
-                        });
-                    }
-                    if let Some(dep) = deps.get(dep) {
-                        visit(dep, deps, order, visiting)?;
-                    }
-                    assert!(visiting.remove(dep));
-                }
-                assert!(order.insert(pkg.name.clone()));
-                Ok(())
-            })
+        for pkg in pkgs {
+            let deps = path.join("deps");
+            let mut deps = self.parse_deps_dir(&deps).with_context(|| {
+                format!("failed to parse dependency directory: {}", deps.display())
+            })?;
+
+            // Perform a simple topological sort which will bail out on cycles
+            // and otherwise determine the order that packages must be added to
+            // this `Resolve`.
+            let mut order = IndexSet::new();
+            let mut visiting = HashSet::new();
+            for dep_pkg in deps.values().chain([&pkg]) {
+                visit(&dep_pkg, &deps, &mut order, &mut visiting)?;
+            }
+
+            // Using the topological ordering insert each package incrementally.
+            // Additionally note that the last item visited here is the root
+            // package, which is the one returned here.
+            let mut last = None;
+            let mut files = Vec::new();
+            let mut pkg = Some(pkg);
+            for name in order {
+                let pkg = deps.remove(&name).unwrap_or_else(|| pkg.take().unwrap());
+                files.extend(pkg.source_files().map(|p| p.to_path_buf()));
+                let pkgid = self.push(pkg)?;
+                last = Some(pkgid);
+            }
+
+            ret.push((last.unwrap(), files));
         }
+
+        Ok(ret)
     }
 
+    // TODO: this is a pretty naive implementation, since for example it doesn't re-use dependencies
+    // between "multi-"packages. Could maybe do better going forward?
```

Comment:
I'm a bit worried about this in that `.insert` below will silently overwrite preexisting packages. Instead I think it'd be best to preserve everything to get errors if necessary.

In refactoring this, what about:

* This method returns `Vec<UnresolvedPackage>` instead of a `BTreeMap`
* The list of packages from the original parse that callers have is appended to this array
* This one big array of `UnresolvedPackage` is top-o sorted with `visit` plus perhaps a `sort_by_key` or similar to generate a sorted array
* The big array is iterated over with `.into_iter()` and pushed iteratively into `Resolve`

Reply:
I went with a similar approach, based on what you suggested in another thread:

- `Vec<UnresolvedPackage>` becomes `UnresolvedPackageGroup`. This stores a `Vec<UnresolvedPackage>`, plus the `SourceMap` for the entire WIT source. This means that `source_map` is no longer a property of `UnresolvedPackage`, which makes sense: multiple packages can potentially share the same source, so copying this around (or dropping it for packages 1..N, as I was erroneously doing in the previous draft) is wasteful.
- There are now two topological sorts that happen: one for top-level packages, and another for the deps. This made the shared `SourceMap` machinery easier: each `dep` gets a `SourceMap`, as does the top-level package, but there is no need for them to be shared beyond that, since each such `SourceMap` must be a discreet set of one or more complete packages. The sort-twice approach is safe, since the packages in the top-level set of files can never be referenced by a `deps` package, so doing separate sorts and then concatenating the results should be fine. This is essentially what we did before with the .chain call, but now we have to sort the top-level packages as well, since those may contain references into one another.